// layer-ui-web/lib-es5/mixins/empty-list.js
/**
 * A helper mixin for Lists that render alternate text in the event that the list is Empty.
 *
 * @class layerUI.mixins.EmptyList
 */
'use strict';


module.exports = {
  properties: {
    /**
     * If the query has no data and is not loading data, this should be true.
     *
     * @property {Boolean} [isEmptyList=false]
     * @readonly
     */
    isEmptyList: {
      value: false,
      set: function set(value) {
        this.nodes.emptyNode.style.display = value ? '' : 'none';
      }
    },

    /**
     * A dom node to render when there are no messages in the list.
     *
     * Could just be a message "Empty Conversation".  Or you can add interactive widgets.
     *
     * @property {HTMLElement} [emptyNode=null]
     */
    emptyNode: {
      set: function set(value) {
        this.nodes.emptyNode.innerHTML = '';
        if (value) this.nodes.emptyNode.appendChild(value);
      }
    }
  },
  methods: {
    onRender: function onRender() {
      this.nodes.emptyNode.style.display = this.isEmptyList ? '' : 'none';
      if (this.emptyNode) this.nodes.emptyNode.appendChild(this.emptyNode);
    },


    /**
     * Call this on any Query change events.
     *
     * @method onRerender
     * @private
     * @param {Event} evt
     */
    onRerender: function onRerender() {
      var evt = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (this.query.isDestroyed) {
        this.isEmptyList = false;
      } else {
        this.isEmptyList = evt.type !== 'reset' && this.query.data.length === 0;
      }
    }
  }
};
