// layer-ui-web/lib-es5/mixins/list.js
/**
 * A List Mixin that provides common list patterns
 *
 * @class layerUI.mixins.List
 * @mixin layerUI.mixins.HasQuery
 */
'use strict';

var _layerWebsdk = require('layer-websdk');

var _layerWebsdk2 = _interopRequireDefault(_layerWebsdk);

var _base = require('../base');

var _component = require('../components/component');

var _hasQuery = require('./has-query');

var _hasQuery2 = _interopRequireDefault(_hasQuery);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }


module.exports = {
  mixins: [_hasQuery2.default],
  properties: {
    /**
     * Lists have some special behaviors; its useful to be able to test if a component is in fact a list.
     *
     * @property {Boolean} [_isList=true]
     * @private
     * @readonly
     */
    _isList: {
      value: true
    },

    /**
     * Set/get state related to whether the Query data is loading data from the server.
     *
     * This is managed by the app, and is updated any time the layer.Query changes state.
     *
     * You could set this as well if you need to indicate some activity outside of the layer.Query:
     *
     * ```
     * widget.isDataLoading = true;
     * ```
     *
     * @property {Boolean} [isDataLoading=false]
     */
    isDataLoading: {},

    /**
     * Any time we are about to render an object, call any provided onRenderListItem function to see if there
     * are nodes to be inserted before/after the User Item.
     *
     * ```javascript
     * userList.onRenderListItem = function(widget, dataArray, index, isTopItemNew) {
     *     var conversation = widget.item;
     *     var priorConversation = dataArray[index - 1];
     *     if (index > 0 && conversation.metadata.category !== priorConversation.metadata.category) {
     *        widget.customNodeAbove = '<div class="my-separator">' + widget.user.metadata.category + '</div>';
     *     }
     * });
     * ```
     *
     * Typical actions on receiving a widget is to set its customNodeAbove and/or customNodeBelow to either a DOM node or an HTML String.
     *
     * @property {Function} [onRenderListItem=null]      Function to call on each rendered item.
     * @property {layer.Root} onRenderListItem.widget    Current user/message/conversation/list-item widget that has been created from the Query.
     * @property {layer.Root[]} onRenderListItem.items   Full set of users/messages/conversations have been/will be rendered
     * @property {Number} onRenderListItem.index         Index of the user/message/conversation in the items array
     * @property {Boolean} onRenderListItem.isTopItemNew If the top item is index 0, and its newly added rather than just affected by changes
     *           around it, this is often useful to know.
     */
    onRenderListItem: {
      type: Function
    },

    /**
     * How many items to page in each time we page the Query.
     *
     * @property {Number} [pageSize=50]
     */
    pageSize: {
      value: 50
    },

    /**
     * A throttler is used to prevent excessive scroll events.
     *
     * This timeout indicates how frequently scroll events are allowed to fire in miliseconds.
     * This value should not need to be tinkered with.
     *
     * @property {Number} [throttlerTimeout=66]
     */
    throttlerTimeout: {
      value: 66
    },

    state: {
      set: function set(newState) {
        Array.prototype.slice.call(this.childNodes).forEach(function (node) {
          node.state = newState;
        });
      }
    },

    /**
     * String, Regular Expression or Function for filtering Conversations.
     *
     * Defaults to filtering by comparing input against things like Conversation.metadata.conversationName, or Identity.displayName, etc.
     * Provide your own Function to change this behavior
     *
     * @property {String|RegEx|Function} [filter='']
     */
    filter: {
      set: function set(value) {
        this._runFilter();
      }
    }
  },
  methods: {
    onCreate: function onCreate() {
      if (!this.id) this.id = _layerWebsdk2.default.Util.generateUUID();
      this.properties.listData = [];
      this.addEventListener('scroll', this._onScroll.bind(this));
      this.onRender();
    },


    /**
     * The _onScroll method throttles calls to the handleScroll method.
     *
     * @method _onScroll
     * @param {Event} evt
     * @private
     */
    _onScroll: function _onScroll(evt) {
      if (this.properties.isSelfScrolling) {
        evt.preventDefault();
      } else {
        this._throttler(this._handleScroll.bind(this));
      }
    },


    /**
     * Simple throttler to avoid too many events while scrolling.
     *
     * Not at this time safe for handling multiple types of events at the same time.
     *
     * @method _throttler
     * @private
     */
    _throttler: function _throttler(callback) {
      var _this = this;

      if (!this.properties.throttleTimeout) {
        this.properties.throttleTimeout = setTimeout(function () {
          _this.properties.throttleTimeout = null;
          callback();
        }, this.throttlerTimeout);
      }
    },


    /**
     * Any time we get a new Query assigned, wire it up.
     *
     * @method _updateQuery
     * @private
     */
    _updateQuery: function _updateQuery() {
      this.query.on('change:property', this._runFilter, this);
    },


    /**
     * If the user scrolls to the bottom of the list, page the Query.
     *
     * @method _handleScroll
     * @private
     */
    _handleScroll: function _handleScroll() {
      if (this.scrollTop >= this.scrollHeight - this.clientHeight - 20 && this.scrollTop > 0) {
        this.query.update({ paginationWindow: this.query.paginationWindow + this.pageSize });
        this.isDataLoading = this.properties.query.isFiring;
      }
    },


    /**
     * Scroll the list to the specified Y position.
     *
     * @method scrollTo
     * @param {Number} position
     */
    scrollTo: function scrollTo(position) {
      if (position === this.scrollTop) return;
      this.scrollTop = position;
    },


    /**
     * Animated scroll to the specified Y position.
     *
     * @method animatedScrollTo
     * @param {Number} position
     * @param {Number} [animateSpeed=200]   Number of miliseconds of animated scrolling; 0 for no animation
     * @param {Function} [animateCallback] Function to call when animation completes
     */
    animatedScrollTo: function animatedScrollTo(position) {
      var _this2 = this;

      var animateSpeed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 200;
      var animateCallback = arguments[2];

      if (this.properties.cancelAnimatedScroll) this.properties.cancelAnimatedScroll();

      var cancel = this.properties.cancelAnimatedScroll = (0, _base.animatedScrollTo)(this, position, animateSpeed, function () {
        if (cancel !== _this2.properties.cancelAnimatedScroll) return;
        _this2.properties.cancelAnimatedScroll = null;
        if (animateCallback) animateCallback();
      });
    },


    /**
     * Scroll to the specified item.
     *
     * Item is assumed to be a layer.Message, layer.Conversation, or whatever the core
     * data set is that is in your list.  Note that this does not load the item from the server;
     * scrolling to an item not in the list will return `false`.
     *
     * @method scrollToItem
     * @param {layer.Root} item
     * @param {Number} [animateSpeed=0]   Number of miliseconds of animated scrolling; 0 for no animation
     * @param {Function} [animateCallback] Function to call when animation completes
     * @return {Boolean}                  Returns true if operation was successful,
     *                                    returns false if the item was not found in the list.
     */
    scrollToItem: function scrollToItem(item) {
      var animateSpeed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      var animateCallback = arguments[2];

      var widget = document.getElementById(this._getItemId(item.id));
      if (!widget) return false;

      var position = widget.offsetTop - this.offsetTop;
      if (!animateSpeed) {
        this.scrollTop = position;
      } else {
        this.animatedScrollTo(position, animateSpeed, animateCallback);
      }

      return true;
    },
    onRender: function onRender() {
      var _this3 = this;

      // Reset the query to initial state by cloning the template
      Array.prototype.slice.call(this.childNodes).forEach(function (node) {
        if (node._isListItem) _this3.removeChild(node);
      });

      // Render any data in the query
      if (this.query && this.query.size) {
        this.onRerender({ type: 'data', data: this.query.data, inRender: true });
      }
    },


    onRerender: {
      mode: _component.registerComponent.MODES.BEFORE,
      value: function value() {
        var evt = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        if (this.query.isDestroyed) {
          this._renderResetData(evt);
        } else {
          this._processQueryEvt(evt);
        }
      }
    },

    /**
     * Generate a document fragment with all the newly added Users.
     *
     * @method _generateFragment
     * @private
     */
    _generateFragment: function _generateFragment(data, fragment) {
      var _this4 = this;

      if (!fragment) fragment = document.createDocumentFragment();
      data.forEach(function (item, index) {
        _this4._generateFragmentItem(item, fragment);
      }, this);
      return fragment;
    },


    /**
     * Generate a unique but consistent DOM ID for each layerUI.mixins.ListItem.
     *
     * @method _getItemId
     * @param {String} itemId
     * @private
     */
    _getItemId: function _getItemId(itemId) {
      return 'list-item-' + this.id + '-' + itemId.replace(/[^a-zA-Z0-9-_]/g, '-').replace(/-+/g, '-');
    },


    /**
     * Generate an list-item for one query result.
     *
     * @method _generateFragmentItem
     * @private
     */
    _generateFragmentItem: function _generateFragmentItem(item, fragment) {
      var itemInstance = item instanceof _layerWebsdk2.default.Root ? item : this.client.getObject(item.id);
      if (itemInstance) {
        var widget = this._generateItem(itemInstance);
        widget.parentComponent = this;
        widget.setAttribute('layer-item-id', item.id.replace(/^layer:\/\/\//, '').replace(/\//g, '_'));
        if (widget) {
          this.onGenerateListItem(widget);
          fragment.appendChild(widget);
        }
      }
    },


    /**
     * MIXIN HOOK: Each time a List Item is generated, call this so that listeners can use this.
     *
     * This is intended for Mixins to hook into; apps wanting to do processing on rendered
     * items should use `onRenderListItem`.
     *
     * @method
     * @param {layerUI.mixins.ListItem} widget
     */
    onGenerateListItem: function onGenerateListItem(widget) {
      // No-op
    },


    /**
     * Find the widgets associated with each affected item and feed it to processAffectedWidgets.
     *
     * @method _gatherAndProcessAffectedItems
     * @private
     */
    _gatherAndProcessAffectedItems: function _gatherAndProcessAffectedItems(affectedItems, isTopItemNew) {
      var _this5 = this;

      if (affectedItems.length) {
        var itemIds = affectedItems.map(function (item) {
          return _this5._getItemId(item.id);
        });
        var affectedWidgets = this.querySelectorAllArray('#' + itemIds.join(', #'));
        this._processAffectedWidgets(affectedWidgets, isTopItemNew);
      }
    },


    /**
     * For all newly added items, as well as items near them,
     * call onRenderListItem and _processAffectedWidgetsCustom to udpate
     * any rendering state needed.
     *
     * widgets are assumed to be sequential within the list.
     *
     * @method _processAffectedWidgets
     * @private
     */
    _processAffectedWidgets: function _processAffectedWidgets(widgets, isTopItemNew) {
      var _this6 = this;

      // Get the index of our first widget within listData
      var firstIndex = void 0;
      for (var i = 0; i < this.properties.listData.length; i++) {
        if (widgets.length && widgets[0].item.id === this.properties.listData[i].id) {
          firstIndex = i;
          break;
        }
      }

      // Do our internal processing of these widgets
      this._processAffectedWidgetsCustom(widgets, firstIndex, isTopItemNew);

      // Allow external processing of the widgets
      widgets.forEach(function (widget, index) {
        if (_this6.properties.onRenderListItem) {
          try {
            _this6.properties.onRenderListItem(widget, _this6.properties.listData, firstIndex + index, isTopItemNew);
          } catch (err) {
            console.error('Error in onRenderListItem for ' + widget.item.id + '; ' + err);
          }
        }
      }, this);
    },


    /**
     * Lists should override this to provide custom behaviors on newly added/affected items.
     *
     * @method _processAffectedWidgetsCustom
     * @private
     * @param {layerUI.mixins.ListItem} widgets
     * @param {Number} firstIndex - Index in the listData array of the first item in the widgets array
     * @param {Boolean} isTopItemNew - If the top item is index 0 and its a new item rather than an "affected" item, this is true.
     */
    _processAffectedWidgetsCustom: function _processAffectedWidgetsCustom(widgets, firstIndex, isTopItemNew) {},


    /**
     * Call this on any Query change events.
     *
     * TODO: This should work on the MessageList which is in Reverse Order
     *
     * @method _processQueryEvt
     * @private
     * @param {Event} evt
     */
    _processQueryEvt: function _processQueryEvt(evt) {
      switch (evt.type) {
        case 'data':
          this._renderPagedData(evt);
          break;
        case 'insert':
          this._renderInsertedData(evt);
          break;
        case 'remove':
          this._renderWithoutRemovedData(evt);
          break;
        case 'reset':
          this._renderResetData(evt);
          break;
        case 'move':
          this._renderMovedData(evt);
      }
    },


    /**
     * The query has been reset of all data, perhaps its now got a new predicate.
     *
     * Clear all data and list state
     *
     * @method _renderResetData
     * @private
     */
    _renderResetData: function _renderResetData(evt) {
      this.properties.listData = [];
      this.scrollTo(0);
      this.onRender();
    },


    /**
     * The query results have had an element move from one position to another.
     *
     * We need to update our list to reflect that change.
     *
     * @method _renderMovedData
     * @private
     */
    _renderMovedData: function _renderMovedData(evt) {
      var oldIndex = evt.fromIndex;
      var newIndex = evt.toIndex;
      var moveNode = this.childNodes[oldIndex];
      this.removeChild(moveNode);
      this.insertBefore(moveNode, this.childNodes[newIndex]);
      if (!evt.inRender) this.onRerender();
    },


    /**
     * Data has been removed from the query; remove that data from our UI.
     *
     * Calls _gatherAndProcessAffectedItems on 3 items prior and 3 items after the removed item.
     *
     * @method _renderWithoutRemovedData
     * @private
     */
    _renderWithoutRemovedData: function _renderWithoutRemovedData(evt) {
      this.properties.listData = [].concat(this.properties.query.data);
      var removeIndex = evt.index;
      var affectedItems = this.properties.listData.slice(Math.max(0, removeIndex - 3), removeIndex + 3);
      var listItem = this.querySelector('#' + this._getItemId(evt.target.id));
      if (listItem) this.removeChild(listItem);

      this._gatherAndProcessAffectedItems(affectedItems, false);
    },


    /**
     * Data has been inserted into the results; insert it into our UI list.
     *
     * @method _renderInsertedData
     * @private
     */
    _renderInsertedData: function _renderInsertedData(evt) {
      this.properties.listData = [].concat(this.properties.query.data);
      var insertIndex = evt.index;
      var affectedItems = this.properties.listData.slice(Math.max(0, insertIndex - 3), insertIndex + 4);
      var fragment = this._generateFragment([evt.target]);
      this.insertBefore(fragment, this.childNodes[insertIndex]);
      this._gatherAndProcessAffectedItems(affectedItems, insertIndex === 0);
    },


    /**
     * A new page of data has been loaded by the query; insert it into our results.
     *
     * @method _renderPagedData
     * @private
     */
    _renderPagedData: function _renderPagedData(evt) {
      var affectedItems = this.properties.listData.slice(this.properties.listData.length - 3, this.properties.listData.length).concat(evt.data);
      this.properties.listData = [].concat(this.properties.query.data);
      var fragment = this._generateFragment(evt.data);

      this.insertBefore(fragment, this.nodes.loadIndicator);

      // isTopItemNew is true if there wasn't any prior data... data length == event length
      this._gatherAndProcessAffectedItems(affectedItems, evt.data.length === this.properties.query.data.length);
      this.isDataLoading = this.properties.query.isFiring;
      if (!evt.inRender) this.onRerender();
    },


    /**
     * Run the filter on all Identity Items.
     *
     * @method _runFilter
     * @private
     */
    _runFilter: function _runFilter() {
      if (!this.filter) {
        this.querySelectorAllArray('.layer-item-filtered').forEach(function (item) {
          return item.removeClass('layer-item-filtered');
        });
      } else {
        for (var i = 0; i < this.childNodes.length; i++) {
          var listItem = this.childNodes[i];
          if (listItem.item instanceof _layerWebsdk2.default.Root) {
            listItem._runFilter(this.filter);
          }
        }
      }
    }
  }
};
