// layer-ui-web/lib-es5/mixins/main-component.js
/**
 * A Mixin for main components (not needed for subcomponents) that provides common properties, shortcuts and code.
 *
 * @class layerUI.mixins.MainComponent
 */
'use strict';

var _layerWebsdk = require('layer-websdk');

var _layerWebsdk2 = _interopRequireDefault(_layerWebsdk);

var _base = require('../base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }


module.exports = {
  properties: {
    /**
     * Is this component a Main Component (high level for use by third party apps).
     *
     * Used by adapters to find components to adapt.
     * @private
     * @readonly
     * @property {Boolean} [_isMainComponent=true]
     */
    _isMainComponent: {
      value: true
    },

    /**
     * An App ID can be provided as a property; this allows the app to find its Client.
     *
     * App IDs are typically provided via:
     *
     * ```
     * layerUI.init(({ appId: myAppId })
     * ```
     *
     * The only time one would use this property
     * is if building an app that used multiple App IDs.
     *
     * @property {String} [appId=""]
     */
    appId: {
      order: 1,
      set: function set(value) {
        var _this = this;

        if (value && value.indexOf('layer:///') === 0) {
          var client = _layerWebsdk2.default.Client.getClient(value);
          if (client) {
            this.client = client;
          } else if (_layerWebsdk2.default.Client.addListenerForNewClient) {
            // Wait for the next client with this appId to be registered and then assign it.
            _layerWebsdk2.default.Client.addListenerForNewClient(function (newClient) {
              return _this.client = newClient;
            }, value);
          } else {
            throw new Error('You must create a layer.Client with your appId before creating this component. Or upgrade to Layer WebSDK 3.2.2 or above.');
          }
        }
      }
    },

    /**
     * The layer.Client can be passed in via the `client` property or the `appId` property.
     *
     * App IDs are typically provided via:
     *
     * ```
     * layerUI.init(({ appId: myAppId })
     * ```
     *
     * The only time one would use this property
     * is if building an app that used multiple Clients.
     *
     * @property {layer.Client} [client=null]
     */
    client: {
      order: 2,
      set: function set(value) {
        var _this2 = this;

        if (value) {
          if (value.telemetryMonitor) {
            value.telemetryMonitor.on('telemetry-environment', function (evt) {
              evt.environment.layer_ui_sdk_version = _base.version;
            });
          }
          value.on('destroy', function (evt) {
            if (evt.target === value) _this2.properties.client = null;
          }, this);
        }
      }
    }
  },
  methods: {
    onCreate: function onCreate() {
      if (_base.settings.appId) this.appId = _base.settings.appId;
      var useSafariCss = navigator.vendor && navigator.vendor.indexOf('Apple') > -1;
      if (useSafariCss) this.classList.add('safari');
    }
  }
};
