// layer-ui-web/lib-es5/utils/files.js
/**
 * This is a utility class which you can use to watch for user dragging and dropping
 * files from their file system into your app as part of a "Send Attached Message" action.
 *
 * ```
 * var dropWatcher = new layerUI.files.DragAndDropFileWatcher({
 *   node: dropBoxNode,
 *   callback: sendAttachment,
 *   allowDocumentDrop: false
 * });
 * function sendAttachment(messageParts) {
 *    currentConversation.createMessage({ parts: messageParts }).send();
 * }
 *
 * // If you finish with this component, call destroy to unsubscribe from all events and remove all pointers
 * dropWatcher.destroy();
 * ```
 *
 * @class layerUI.utils.files.DragAndDropFileWatcher
 * @param {Object} options
 * @param {HTMLElement|String} options.node - The dom node (or dom node ID) to watch for files/file-drag events
 * @param {Function} options.callback - The function to call when a file is dropped
 * @param {layer.MessagePart[]} options.callback.parts - The MessageParts representing the dropped files, which you can modify and send.
 * @param {Boolean} [options.allowDocumentDrop=false] - By default, this utility adds an event handler to prevent the browser from navigating away from your
 *         app to view a file dropped in some other part of your app. If you need to handle this event yourself, set this to true.
 */
'use strict';

var _loadImage = require('blueimp-load-image/js/load-image');

var _loadImage2 = _interopRequireDefault(_loadImage);

require('blueimp-load-image/js/load-image-orientation');

require('blueimp-load-image/js/load-image-meta');

require('blueimp-load-image/js/load-image-exif');

var _layerWebsdk = require('layer-websdk');

var _layerWebsdk2 = _interopRequireDefault(_layerWebsdk);

var _base = require('../base');

var _base2 = _interopRequireDefault(_base);

var _sizing = require('./sizing');

var _sizing2 = _interopRequireDefault(_sizing);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Files = {};
module.exports = Files;

window.loadImage = _loadImage2.default;


Files.DragAndDropFileWatcher = function DragAndDropFileWatcher(options) {
  this.node = typeof options.node === 'string' ? document.getElementById(options.node) : options.node;
  this.callback = options.callback;
  this.allowDocumentDrop = Boolean(options.allowDocumentDrop);

  this.onDragOverBound = this.onDragOver.bind(this);
  this.onDragEndBound = this.onDragEnd.bind(this);
  this.onFileDropBound = this.onFileDrop.bind(this);
  this.ignoreDropBound = this.ignoreDrop.bind(this);

  // Tells the browser that we *can* drop on this target
  this.node.addEventListener('dragover', this.onDragOverBound, false);
  this.node.addEventListener('dragenter', this.onDragOverBound, false);

  this.node.addEventListener('dragend', this.onDragEndBound, false);
  this.node.addEventListener('dragleave', this.onDragEndBound, false);

  this.node.addEventListener('drop', this.onFileDropBound, false);

  if (!this.allowDocumentDrop) {
    document.addEventListener('drop', this.ignoreDropBound, false);
    document.addEventListener('dragenter', this.ignoreDropBound, false);
    document.addEventListener('dragover', this.ignoreDropBound, false);
  }
};

/**
 * Destroy this component, in particular, remove any handling of any events.
 *
 * @method
 */
Files.DragAndDropFileWatcher.prototype.destroy = function destroy() {
  this.node.removeEventListener('dragover', this.onDragOverBound, false);
  this.node.removeEventListener('dragenter', this.onDragOverBound, false);

  this.node.removeEventListener('dragend', this.onDragEndBound, false);
  this.node.removeEventListener('dragleave', this.onDragEndBound, false);

  this.node.removeEventListener('drop', this.onFileDropBound, false);

  if (!this.allowDocumentDrop) {
    document.removeEventListener('drop', this.ignoreDropBound, false);
    document.removeEventListener('dragenter', this.ignoreDropBound, false);
    document.removeEventListener('dragover', this.ignoreDropBound, false);
  }
  delete this.node;
  delete this.callback;
};

/**
 * Whatever it is that the browser wants to do by default with this file,
 * prevent it.  Why? Well, one of the more annoying thing it may do
 * is navigate away from your app to show this file.
 *
 * @method
 * @private
 */
Files.DragAndDropFileWatcher.prototype.ignoreDrop = function ignoreDrop(evt) {
  if (evt.preventDefault) {
    evt.preventDefault();
    evt.stopPropagation();
  }
  return false;
};

/**
 * On hovering with a file, add a css class
 *
 * @method
 * @private
 */
Files.DragAndDropFileWatcher.prototype.onDragOver = function onDragOver(evt) {
  this.node.classList.add('layer-file-drag-and-drop-hover');
  evt.preventDefault();
  return false;
};

/**
 * On un-hovering with a file, remove a css class
 *
 * @method
 * @private
 */
Files.DragAndDropFileWatcher.prototype.onDragEnd = function onDragEnd(evt) {
  this.node.classList.remove('layer-file-drag-and-drop-hover');
};

/**
 * On file drop, generate an array of Message Parts and pass it on to app.
 *
 * @method
 * @private
 */
Files.DragAndDropFileWatcher.prototype.onFileDrop = function onFileDrop(evt) {
  this.onDragEnd();

  // stops the browser from redirecting off to the image.
  if (evt.preventDefault) {
    evt.preventDefault();
    evt.stopPropagation();
  }

  var dt = evt.dataTransfer;
  var parts = Array.prototype.map.call(dt.files, function (file) {
    return new _layerWebsdk2.default.MessagePart(file);
  });

  Files.processAttachments(parts, this.callback);
  return false;
};

/**
 * Adds data to your message.
 *
 * Given an array of Message Parts, determines if it needs to generate image or video
 * previews and metadata message parts before calling your callback.
 *
 * @method processAttachments
 * @param {layer.MessagePart[]} parts    Input MessagParts, presumably an array of one element
 * @param {Function} callback            Callback on completion; may be called synchronously
 * @param {layer.MessagePart[]} callback.parts  The MessageParts to send in your Message
 */
Files.processAttachments = function processAttachments(parts, callback) {
  // TODO: Need a way to register additional handlers; currently relies on the callback for additional handling.
  parts.forEach(function (part) {
    if (['image/gif', 'image/png', 'image/jpeg'].indexOf(part.mimeType) !== -1) {
      Files.generateImageMessageParts(part, callback);
    } else if (part.mimeType === 'video/mp4') {
      Files.generateVideoMessageParts(part, callback);
    } else if (callback) {
      callback([part]);
    }
  });
};

Files.generateImageMessageParts = function generateImageMessageParts(part, callback) {
  // First part is the original image; the rest of the code is for generating the other 2 parts of the 3 part Image
  var parts = [part];
  var orientation = 0;

  // STEP 1: Determine the correct orientation for the image
  _loadImage2.default.parseMetaData(part.body, onParseMetadata);

  function onParseMetadata(data) {
    var options = {
      canvas: true
    };

    if (data.imageHead && data.exif) {
      orientation = options.orientation = data.exif[0x0112] || orientation;
    }

    // STEP 2: Write the image to a canvas with the specified orientation
    (0, _loadImage2.default)(part.body, onWriteImage, options);
  }

  function onWriteImage(srcCanvas) {
    // STEP 3: Scale the image down to Preview Size
    var originalSize = {
      width: srcCanvas.width,
      height: srcCanvas.height
    };

    var size = (0, _sizing2.default)(originalSize, _base2.default.settings.maxSizes);
    var canvas = document.createElement('canvas');
    canvas.width = size.width;
    canvas.height = size.height;
    var context = canvas.getContext('2d');

    // context.scale(size.width, size.height);
    context.fillStyle = context.strokeStyle = 'white';
    context.fillRect(0, 0, size.width, size.height);
    context.drawImage(srcCanvas, 0, 0, size.width, size.height);

    // STEP 4: Turn the canvas into a jpeg image for our Preview Image
    var binStr = atob(canvas.toDataURL('image/jpeg').split(',')[1]);
    var len = binStr.length;
    var arr = new Uint8Array(len);

    for (var i = 0; i < len; i++) {
      arr[i] = binStr.charCodeAt(i);
    }
    var blob = new Blob([arr], { type: 'image/jpeg' });

    // STEP 5: Create our Preview Message Part
    parts.push(new _layerWebsdk2.default.MessagePart({
      body: blob,
      mimeType: 'image/jpeg+preview'
    }));

    // STEP 6: Create the Metadata Message Part
    parts.push(new _layerWebsdk2.default.MessagePart({
      mimeType: 'application/json+imageSize',
      body: JSON.stringify({
        orientation: orientation,
        width: originalSize.width,
        height: originalSize.height,
        previewWidth: canvas.width,
        previewHeight: canvas.height
      })
    }));
    callback(parts);
  }
};

Files.generateVideoMessageParts = function generateVideoMessageParts(part, callback) {
  var parts = [part];
  var video = document.createElement('video');

  video.addEventListener('loadedmetadata', function () {
    var originalSize = {
      width: video.videoWidth,
      height: video.videoHeight
    };

    var size = (0, _sizing2.default)(originalSize, _base2.default.settings.maxSizes);

    var canvas = document.createElement('canvas');
    canvas.width = size.width;
    canvas.height = size.height;
    var context = canvas.getContext('2d');
    context.drawImage(video, 0, 0, canvas.width, canvas.height);

    var binStr = atob(canvas.toDataURL('image/jpeg').split(',')[1]);
    var len = binStr.length;
    var arr = new Uint8Array(len);

    for (var i = 0; i < len; i++) {
      arr[i] = binStr.charCodeAt(i);
    }

    parts.push(new _layerWebsdk2.default.MessagePart({
      body: new Blob([arr], { type: 'image/jpeg' }),
      mimeType: 'image/jpeg+preview'
    }));

    parts.push(new _layerWebsdk2.default.MessagePart({
      mimeType: 'application/json+imageSize',
      body: '{"orientation":0, "width":' + originalSize.width + ', "height":' + originalSize.height + '}'
    }));

    if (callback) callback(parts);
  });

  video.src = URL.createObjectURL(part.body);
};
