// layer-ui-web/lib-es5/adapters/react.js
/**
 * Call this function to initialize all of the react components needed to handle the Layer UI for Web widgets.
 *
 * Before using this, please note that layerUI.init() must be called prior to calling layerUI.adapters.react().
 *
 * Initialize with:
 *
 * ```
 * import React from 'react';
 * import ReactDom from 'react-dom';
 * const { ConversationPanel, ConversationList, UserList, Notifier } = layerUI.adapters.react(React, ReactDom);
 * ```
 *
 * Calling this will expose the following React Components:
 *
 * * ConversationPanel: A wrapper around a layerUI.components.ConversationPanel
 * * ConversationsList: A wrapper around a layerUI.components.ConversationsListPanel
 * * IdentitiesList: A wrapper around a layerUI.components.IdentitiesListPanel
 * * Notifier: A wrapper around a layerUI.components.misc.Notifier
 * * SendButton: A wrapper around a layerUI.components.subcomponents.SendButton
 * * FileUploadButton: A wrapper around a layerUI.components.subcomponents.FileUploadButton
 *
 * You can then use:
 *
 * ```
 * render() {
 *    return <ConversationList
 *      composeButtons={SendButton, FileUploadButton}
 *      onConversationSelected={this.mySelectHandler}></ConversationList>
 * }
 * ```
 *
 * To insure that LayerUI.init() is called before layerUI.adapters.react(), and each is only called once, we
 * recommend puttings this code in its own module:
 *
 * ```
 * import React, { Component, PropTypes } from 'react';
 * import ReactDom from 'react-dom';
 * import Layer from 'layer-websdk';
 * import * as LayerUI from 'layer-ui-web';
 *
 * LayerUI.init({
 *   appId: 'layer:///apps/staging/my-app-id',
 *   layer: Layer
 * });
 * const LayerUIWidgets = LayerUI.adapters.react(React, ReactDom);
 * module.exports = LayerUIWidgets;
 * ```
 *
 * Now anywhere you need access to the LayerUIWidgets library can import this module and expect everything to
 * evaluate at the correct time, correct order, and only evaluate once.
 *
 * @class layerUI.adapters.react
 * @singleton
 * @param {Object} React - Pass in the reactJS library
 * @param {Object} ReactDom - Pass in the ReactDom library
 */
'use strict';

var _base = require('../base');

var _base2 = _interopRequireDefault(_base);

var _layerWebsdk = require('layer-websdk');

var _layerWebsdk2 = _interopRequireDefault(_layerWebsdk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }


var libraryResult = void 0;
function initReact(React, ReactDom) {
  if (libraryResult) return libraryResult;
  libraryResult = {};

  // Gather all UI Components flagged as Main Components; other components don't require special React Components for direct use.
  Object.keys(_base2.default.components).filter(function (componentName) {
    var component = _base2.default.components[componentName];
    return component.properties.filter(function (prop) {
      return prop.propertyName === '_isMainComponent';
    }).length;
  }).forEach(function (componentName) {
    var component = _base2.default.components[componentName];

    // Get the camel case Component name
    var className = (componentName.substring(0, 1).toUpperCase() + componentName.substring(1).replace(/-(.)/g, function (str, value) {
      return value.toUpperCase();
    })).replace(/^Layer/, '');

    libraryResult[className] = React.createClass({
      /**
       * On mounting, copy in all properties, and optionally setup a Query.
       *
       * Delay added to prevent Webcomponents property setters from being blown away in safari and firefox
       */
      componentDidMount: function componentDidMount() {
        var _this = this;

        // Get the properties/attributes that match those used in this.props
        var props = component.properties.filter(function (property) {
          return _this.props[property.propertyName] || _this.props[property.attributeName];
        });

        // Set the webcomponent properties
        props.forEach(function (propDef) {
          var value = propDef.propertyName in _this.props ? _this.props[propDef.propertyName] : _this.props[propDef.attributeName];
          if (propDef.type === HTMLElement && value) {
            value = _this.handleReactDom(propDef, value);
          }
          _this.node[propDef.propertyName] = value;
        });

        // Browsers running the polyfil may not yet have initialized the component at this point.
        // Force them to be initialized so that by the time the parent component's didComponentMount
        // is called, this will be an initialized widget.
        if (!this.node._onAfterCreate) {
          var evt = document.createEvent('CustomEvent');
          evt.initCustomEvent('HTMLImportsLoaded', true, true, null);
          document.dispatchEvent(evt);
        }
        this.node._onAfterCreate();
      },


      /**
       * Copy all properties into the dom node, but never let React recreate this widget.
       */
      shouldComponentUpdate: function shouldComponentUpdate(nextProps) {
        var _this2 = this;

        // Get the properties/attributes that match those used in this.props
        var props = component.properties.filter(function (property) {
          return _this2.props[property.propertyName] || _this2.props[property.attributeName];
        });

        // Set the webcomponent properties if they have changed
        props.forEach(function (propDef) {
          var name = propDef.propertyName in _this2.props ? propDef.propertyName : propDef.attributeName;
          var value = nextProps[name];
          if (propDef.type === HTMLElement && value) {
            value = _this2.handleReactDom(propDef, value);
          }

          if (value !== _this2.props[name]) {
            _this2.node[propDef.propertyName] = value;
          }
        }, this);
        return false;
      },
      handleReactDom: function handleReactDom(propDef, value) {
        if (!this.layerUIGeneratedNodes) this.layerUIGeneratedNodes = {};

        if (Array.isArray(value)) {
          var array = [];
          if (!this.layerUIGeneratedNodes[propDef.propertyName]) {
            this.layerUIGeneratedNodes[propDef.propertyName] = array;
          }
          array.length = value.length;
          value.forEach(function (item, index) {
            if (item.tagName) {
              array[index] = item;
            } else {
              var node = array[index] || document.createElement('div');
              ReactDom.render(typeof item === 'function' ? React.createElement(item) : item, node);
              array[index] = node;
            }
          });
        } else if (value.tagName === undefined) {
          if (!this.layerUIGeneratedNodes[propDef.propertyName]) {
            this.layerUIGeneratedNodes[propDef.propertyName] = document.createElement('div');
          }
          ReactDom.render(value, this.layerUIGeneratedNodes[propDef.propertyName]);
        }
        return this.layerUIGeneratedNodes[propDef.propertyName];
      },
      render: function render() {
        var _this3 = this;

        return React.createElement(componentName, {
          ref: function ref(node) {
            _this3.node = node;
          },
          id: this.props.id
        });
      }
    });
  });
  return libraryResult;
}

module.exports = initReact;
_base2.default.addAdapter('react', initReact);
